src/Features/Core/Portable/UseSystemHashCode/Analyzer.cs

```diff
@@ -0,0 +1,352 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.UseSystemHashCode
+{
+    /// <summary>
+    /// Helper code to support both <see cref="UseSystemHashCodeCodeFixProvider"/> and
+    /// <see cref="UseSystemHashCodeDiagnosticAnalyzer"/>.
+    /// </summary>
+    internal struct Analyzer
+    {
+        private readonly Compilation _compilation;
+        private readonly IMethodSymbol _objectGetHashCodeMethod;
+        private readonly INamedTypeSymbol _equalityComparerTypeOpt;
+        public readonly INamedTypeSymbol SystemHashCodeType;
+
+        public Analyzer(Compilation compilation)
+        {
+            _compilation = compilation;
+
+            var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+            _objectGetHashCodeMethod = objectType?.GetMembers(nameof(GetHashCode)).FirstOrDefault() as IMethodSymbol;
+            _equalityComparerTypeOpt = compilation.GetTypeByMetadataName(typeof(EqualityComparer<>).FullName);
+            SystemHashCodeType = compilation.GetTypeByMetadataName("System.HashCode");
+        }
+
+        public bool CanAnalyze()
+            => SystemHashCodeType != null && _objectGetHashCodeMethod != null;
+
+        /// <summary>
+        /// Analyzes the containing <c>GetHashCode</c> method to determine which fields and
+        /// properties were combined to form a hash code for this type.
+        /// </summary>
+        public ImmutableArray<ISymbol> GetHashedMembers(ISymbol owningSymbol, IOperation operation)
+        {
+            Debug.Assert(CanAnalyze());
+
+            // Capture a copy of `this` as we're in a struct and need to access it from local
+            // functions in this method.
+            var _this = this;
+
+            if (!(operation is IBlockOperation blockOperation))
+            {
+                return default;
+            }
+
+            // Owning symbol has to be an override of Object.GetHashCode.
+            if (!(owningSymbol is IMethodSymbol method) ||
+                method.Name != nameof(GetHashCode) ||
+                method.Locations.Length != 1 ||
+                method.DeclaringSyntaxReferences.Length != 1 ||
+                !method.Locations[0].IsInSource ||
+                !OverridesSystemObject(method))
```

Comment:
Might be easier to read in the positive using a local variable etc? Or moving three of the checks into property patterns, e.g. `IMethodSymbol { Name: nameof(GetHashCode), ... }`

Reply:
done.